server/public/model/config.go

```diff
@@ -4328,49 +4328,124 @@ func (o *Config) GetSanitizeOptions() map[string]bool {
 	return options
 }
 
-func (o *Config) Sanitize() {
-	if o.LdapSettings.BindPassword != nil && *o.LdapSettings.BindPassword != "" {
-		*o.LdapSettings.BindPassword = FakeSetting
+type SanitizeConfigRules struct {
+	Contains   []string
+	StartsWith []string
+	EndsWith   []string
+	Whitelist  []string
+}
+
+func sanitizeConfigMap(config map[string]interface{}, rules SanitizeConfigRules, currentPath string) *AppError {
+	for key, value := range config {
+		// Construct the new path
+		newPath := key
+		if currentPath != "" {
+			newPath = currentPath + "." + key
+		}
+
+		switch v := value.(type) {
+		case map[string]interface{}:
+			if err := sanitizeConfigMap(v, rules, newPath); err != nil {
+				return err
+			}
+		case []interface{}:
+			for i, item := range v {
+				if itemMap, ok := item.(map[string]interface{}); ok {
+					if err := sanitizeConfigMap(itemMap, rules, newPath); err != nil {
+						return err
+					}
+				} else if str, ok := item.(string); ok {
+					if shouldSanitize(newPath, key, rules) && str != "" {
+						v[i] = FakeSetting
+					}
+				}
+			}
+		case string:
+			if shouldSanitize(newPath, key, rules) && v != "" {
+				config[key] = FakeSetting
+			}
+		default:
+			continue
+		}
 	}
+	return nil
+}
 
-	if o.FileSettings.PublicLinkSalt != nil {
-		*o.FileSettings.PublicLinkSalt = FakeSetting
+func shouldSanitize(path string, key string, rules SanitizeConfigRules) bool {
+	// Check if the path is in the whitelist
+	for _, whitelistedPath := range rules.Whitelist {
+		if strings.EqualFold(whitelistedPath, path) {
+			return false
+		}
+	}
+	lowerKey := strings.ToLower(key)
+	for _, substring := range rules.Contains {
+		if strings.Contains(lowerKey, strings.ToLower(substring)) {
+			return true
+		}
 	}
+	for _, prefix := range rules.StartsWith {
+		if strings.HasPrefix(lowerKey, strings.ToLower(prefix)) {
+			return true
+		}
+	}
+	for _, suffix := range rules.EndsWith {
+		if strings.HasSuffix(lowerKey, strings.ToLower(suffix)) {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
Wondering if we should go with explicitly setting sanitization via [struct](https://go.dev/ref/spec#Struct_types) tags? I think this makes sense while trying to automate the sanitization but it's also a "magical" way of handling this. 

I think this sanitization should be explicitly set or explicitly ignored. I know this sounds weird but it's intuitive. So My suggestion would be:
Add a new sturct tag to explicitly set sanitization
Add a new [vet checker](https://github.com/mattermost/mattermost-govet) to check possible fields such as `"password", "secret", "key", "keyid", "email"` etc has explicit sanitization or `//nolint` comments.

@pvev @hanzei WDYT?

Reply:
I like the idea of using struct tags for the mattermost config fields. That makes a lot of sense given that the config struct is in this code base. 

But for plugins, we need a go struct that we can fetch. The plugin passes a JSON file containing the config. By adding a new field (e.g. `confidential`) we allow plugins to declare themself, what config values need to be obfuscated in the Support Packet. I already wrote a proposal in https://mattermost.atlassian.net/browse/MM-57194, a ticket that was opened before by @jlandells.

Reply:
@isacikgoz we could keep the struct tags as a long term goal since it would imply as @hanzei mentioned modify all supported plugins and release new versions of those so we can rely on the tags for sanitization. As a short term solution this PR can work, also we could potentially add a similar solution in the admin console (front-end side) for obfuscating plugins information that could be considered sensitive.

Reply:
@hanzei @isacikgoz do you have more questions/concerns around this PR?

Reply:
I wonder if we can get the best of both PRs.


https://github.com/mattermost/mattermost/pull/27986 requires plugin devs to update their plugins.  Until a plugin dev opts into the new secret field that my PR introduces, the "auto-secret detection" introduced in this PR is a great workaround. 

Can we modify the logic in this PR to only apply to plugins that don't use https://github.com/mattermost/mattermost/pull/27986?